src/Perl6/bootstrap.c/BOOTSTRAP.nqp

```diff
@@ -1404,6 +1404,31 @@ class ContainerDescriptor::VivifyHash does ContainerDescriptor::Whence {
         $array.BIND-KEY($!key, $scalar);
     }
 }
+# Attributes that are either required or have a default need us to detect if
+# they have been initialized. We do this by starting them out with a descriptor
+# that indicates they are uninitialized, and then swapping it out for a the
+# underlying one upon assignment.
+class ContainerDescriptor::UninitializedAttribute {
+    has $!next-descriptor;
+
+    method new($next) {
+        my $obj := nqp::create(self);
+        nqp::bindattr($obj, ContainerDescriptor::UninitializedAttribute,
+            '$!next-descriptor', $next);
+        $obj
+    }
+
+    method next() {
+        my $next := $!next-descriptor;
+        nqp::isconcrete($next)
+            ?? $next
+            !! ($!next-descriptor := nqp::gethllsym('Raku', 'default_cont_spec'))
+    }
+    method name() { self.next.name }
+    method of() { self.next.of }
+    method default() { self.next.default }
+    method dynamic() { self.next.dynamic }
+}
```

Comment:
Would this need methods is_generic, is_default_generic, instantiate_generic as well? Or why doesn't it?

Reply:
At first I thought so, but then realized that the timing of things is that generic instantiation happens before we get to wrapping the underlying descriptor up. I can't quite convince myself that we'd never end up trying to instantiate it, though, so I guess better to have those methods...I'll add them.